core/storage/exploration/gae_models.py

```diff
@@ -515,3 +516,93 @@ def get_recently_published(cls, limit):
+    change in future.
+
+    The key of each instance if combination of exploration id and version.
+    """
+
+    # The exploration id whose states are mapped.
+    exploration_id = ndb.StringProperty(indexed=True, required=True)
+
+    # The version of the exploration.
+    exploration_version = ndb.IntegerProperty(required=True)
+
+    # A dict which maps each state name to a unique id.
+    state_name_to_ids = ndb.JsonProperty(required=True)
+
+    # Latest state id that has been assigned to any of the states in any of
+    # of the  versions of given exploration. New state IDs should be assigned
+    # from this value + 1.
+    latest_state_id_used = ndb.IntegerProperty(required=True)
+
+    @classmethod
+    def _generate_instance_id(cls, exp_id, exp_version):
+        """Generates ID of the state id mapping model instance.
+
+        Args:
+            exp_id: str. The exploration id that whose states are mapped.
+            exp_version: int. The version of exploration.
+        """
+        return '%s.%d' % (exp_id, exp_version)
+
+    @classmethod
+    def create(
+            cls, exp_id, exp_version, state_name_to_ids, latest_state_id_used):
+        """Creates a new instance of state id mapping model.
+
+        Args:
+            exp_id: str. The exploration id whose states are mapped.
+            exp_version: int. The version of that exploration.
+            state_name_to_ids: dict. A dict storing state name to ids mapping.
+            latest_state_id_used: int. Latest state id that has been assigned to
+                any of the states in of the versions of given exploration.
+
+        Returns:
+            StateIdMappingModel. Instance of the state id mapping model.
+        """
+        instance_id = cls._generate_instance_id(exp_id, exp_version)
+        if cls.get_by_id(instance_id):
+            raise Exception(
+                'State id mapping model already exists for given exploration'
+                ' version')
+        model = cls(
+            id=instance_id, exploration_id=exp_id,
+            exploration_version=exp_version,
+            state_name_to_ids=state_name_to_ids,
+            latest_state_id_used=latest_state_id_used)
+        model.put()
+
+        return model
+
+    @classmethod
+    def get_state_id_mapping_model(cls, exp_id, exp_version, strict=False):
+        """Retrieve state id mapping model from the datastore.
+
+        Args:
+            exp_id: str. The exploration id.
+            exp_version: int. The exploration version.
+            strict: bool. Whether to raise an error if no StateIdMappingModel
+                entry is found for the given exploration id and version.
+
+        Returns:
+            StateIdMappingModel. The model retrieved from the datastore.
+        """
+        instance_id = cls._generate_instance_id(exp_id, exp_version)
+        instance = cls.get(instance_id, strict)
+        return instance
+
+    def get_all_state_name_ids(self):
```

Comment:
Drop this method; do it in the domain object if needed. The model class should just be used for storage and be very simple.

Reply:
Done. Dropped it.